### PR TITLE
Output streams: Improved error handling and doc

### PR DESCRIPTION
**Pull request by hrittich**
- Added a proper exception when writing to output stream fails.
- Added to the documentation that output streams should be opened in
  binary mode.

#153

### CODE REVIEW

File: cairo/surface.c [hunk 1]
```diff
@@ -126,12 +126,26 @@ static cairo_status_t
...
+    PyErr_Fetch(&ptype, &pvalue, &ptraceback);
+
+    PyObject *error_string
+      = PyUnicode_FromFormat("Cairo: Exception occured while writing to output stream: %U",
+                             pvalue);
+    Py_XDECREF(ptype);
+    Py_XDECREF(pvalue);
+    Py_XDECREF(ptraceback);
+
+    PyErr_SetObject(PyExc_Exception, error_string);
```
Comment (lazka) [inline cairo/surface.c]: This will lead to the exception getting raised in unrelated code once things return to the interpreter.

File: cairo/surface.c [hunk 2]
```diff
@@ -305,7 +319,7 @@ surface_set_device_offset (PycairoSurface *o, PyObject *args) {
   double x_offset, y_offset;
 
   if (!PyArg_ParseTuple (args, "dd:Surface.set_device_offset",
-			 &x_offset, &y_offset))
+                         &x_offset, &y_offset))
```
Comment (lazka) [inline cairo/surface.c]: What's with all the whitespace changes?

File: docs/reference/surfaces.rst [hunk 3]
```diff
@@ -539,7 +539,7 @@ multi-page vector surface backend.
 
 .. class:: PDFSurface(fobj, width_in_points, height_in_points)
 
-   :param fobj: a filename or writable file object. None may be used to specify no output. This will generate a *PDFSurface* that may be queried and used as a source, without generating a temporary file.
+   :param fobj: a filename or writable file object. The file object *must* be opened in binary mode. None may be used to specify no output. This will generate a *PDFSurface* that may be queried and used as a source, without generating a temporary file.
```
Comment (lazka) [inline docs/reference/surfaces.rst]: Looks good :+1:

File: cairo/surface.c (hunk 1 above)
Comment (hrittich) [reply cairo/surface.c]: What do you mean by unrelated code? As soon as a Cairo function tries to write to the stream and fails, it returns because CAIRO_STATUS_WRITE_ERROR is returned. Then Pycairo_Check_Status runs and sees that an exception was raised. Hence, the function that caused the error to trigger will also raise the exception.

File: cairo/surface.c (hunk 2 above)
Comment (hrittich) [reply cairo/surface.c]: There was a mix of tabs and spaces used for indentation, which I assumed where done by accident.

File: cairo/surface.c (hunk 1 above)
Comment (lazka) [reply cairo/surface.c]: After PyGILState_Release() any Python code can run (in another thread for example) and will fail because an exception is set.

Comment (hrittich) [reply cairo/surface.c]: I see. I was not aware of this multi-threading issue. If I find some spare time, I will try to implement a way to properly pass the error message around, if you do not mind.

**Comment by lazka** [comment]
Two things:

1) I would merge the docs changes if you split them out into a new PR.
2) For the bytes/text error case I'm wondering if we could just try to write an empty byte string to the passed in file. This would result in the PDFSurface() constructor failing early with the right error.

**Comment by lazka** [comment]
I've opened #157 for this.

**Comment by lazka** [comment]
>  For the bytes/text error case I'm wondering if we could just try to write an empty byte string to the passed in file. This would result in the PDFSurface() constructor failing early with the right error.

I've implemented this now in #205
